### DISCUSSION TITLE
tilt sensor measured angle changes while no movement of technic hub

### DISCUSSION THREAD

**Discussion by vanczakp** [pybricks/support#1861] [Q&A] [answered]
Hello all, I wanted to create the weight overload function to my Liebherr crane and got a similar but different issue.
While the crane was stable the tilt of hub was around 9deg and stable (no rubbers, jut at the end position). Then I moved the hub, then released to initial position where the tilt was 9deg. Tilt value started to decrease, and after one minute it was around 7.5 only. Same happened when I wanted to have a program which keeps a tank's turntable in position regardless where it is moving. I was told it is due to the sensor re calibrates itself as it heating. Can you advise how to eleminate this issue?

**Comment by laurensvalk** [comment]
If you want to measure `tilt`, it would be best to use the `tilt()` method as the name implies.

If it is reporting a decimal value of 7.5 then it sounds like you are using a different method?

**Reply by vanczakp** [reply]
I was surprised it has like 10digit accuracy because few months ago and few firmware update earlier the same command gave me integer only.
I use beta block coding only. Used the get "hub name" rotation y+ output block in a repeat cycle with print of value and wait of 500ms.
Firmware update was yesterday

**Reply by laurensvalk** [reply]
Right. So it would be better to use `tilt()` to measure tilt.

If you use `rotation()`, then 3-dimensional movements all affect each other. There is also some drift. You can use this in fast moving applications.

For things that are almost stationary, like cranes, tilt is usually the simpler option.

**Reply by vanczakp** [reply]
Thanks. That woud be the pitch, right?

Anyway this is still half solution, its okay for weight overload, but I have an another big project, automatize a lifle like lr13500. For that, to keep the supestructure always on levele, I need more precise measurment than an integer, i need like 1/10 deg or 1/100 accuracy and I definetly want to avoid measuring the accelerations. I want to measure the angle of main and derrick boom as well because counting the cable winch rotation is overkill and unstable.
So the question still applies: do you have plan to eleminate the drift? When it happend the crane was laying on a shelft and the turntable was turned off so it was really a rotation around one axis.

**Reply by laurensvalk** [reply, marked as answer]
Pitch or roll, depending on which axis you're tilting along. I'm open to a feature request to return tilt as floating point values.

> I definetly want to avoid measuring the accelerations

This is ultimately what the `tilt()` method is doing.

**Reply by Debenben** [reply]
I would recommend:
* Use `acceleration` and calculate the angles yourself, that way you are not affected by gyro drift like `rotation()` and you are not limited to integer precision like `tilt()` 
* Calibrate the accelerometer with an offset e.g. somthing like https://github.com/pybricks/support/issues/943#issuecomment-1510481180
* Take an average of at least 10 accelerometer readings

The above should get you to 1/10 deg accuracy without problem, I checked this by attaching laser pointers to a technic hub.

* In case you need more precision you could calibrate it more precisely, e.g. like described here https://www.nxp.com/docs/en/application-note/AN4399.pdf
* In case you need faster reaction time or want to compensate for different temperatures or things like that and you are not afraid of writing some C functions and compiling your own firmware, you should be able to e.g. set a low-pass filter on the accelerometer on a hardware level. If I am not mistaken you can also add a calibration offset here https://github.com/pybricks/pybricks-micropython/blob/master/lib/pbio/drv/imu/imu_lsm6ds3tr_c_stm32.c#L430 with something like
```c
uint32_t pbdrv_imu_get_acceleration_offset(pbdrv_imu_dev_t *imu_dev) {
    uint32_t offset = 0;
    struct pt pt;
    PT_INIT(&pt);
    lsm6ds3tr_c_xl_usr_offset_get(&pt, &imu_dev->ctx, (uint8_t*)&offset);
    lsm6ds3tr_c_xl_offset_weight_get(&pt, &imu_dev->ctx, (lsm6ds3tr_c_usr_off_w_t*)((uint8_t*)&offset) + 3);
    return offset;
}

void pbdrv_imu_set_acceleration_offset(pbdrv_imu_dev_t *imu_dev, uint32_t offset) {
    struct pt pt;
    PT_INIT(&pt);
    lsm6ds3tr_c_xl_offset_weight_set(&pt, &imu_dev->ctx, (lsm6ds3tr_c_usr_off_w_t)(offset & 0x01000000));
    lsm6ds3tr_c_xl_usr_offset_set(&pt, &imu_dev->ctx, (uint8_t*)&offset);
}
```
I have not found motivation to test this further because I do not need that much accuracy, but I would be interested to see the results and help if someone else wants to.

**Reply by vanczakp** [reply]
> Pitch or roll, depending on which axis you're tilting along. I'm open to a feature request to return tilt as floating point values.
> 
> > I definetly want to avoid measuring the accelerations
> 
> This is ultimately what the `tilt()` method is doing.

Thanks, that would be great if you could do it!


> I would recommend:
> 
> * Use `acceleration` and calculate the angles yourself, that way you are not affected by gyro drift like `rotation()` and you are not limited to integer precision like `tilt()`
> * Calibrate the accelerometer with an offset e.g. somthing like [[Feature] User IMU calibration ideas #943 (comment)](https://github.com/pybricks/support/issues/943#issuecomment-1510481180)
> * Take an average of at least 10 accelerometer readings
> 
> The above should get you to 1/10 deg accuracy without problem, I checked this by attaching laser pointers to a technic hub.
> 
> * In case you need more precision you could calibrate it more precisely, e.g. like described here https://www.nxp.com/docs/en/application-note/AN4399.pdf
> * In case you need faster reaction time or want to compensate for different temperatures or things like that and you are not afraid of writing some C functions and compiling your own firmware, you should be able to e.g. set a low-pass filter on the accelerometer on a hardware level. If I am not mistaken you can also add a calibration offset here https://github.com/pybricks/pybricks-micropython/blob/master/lib/pbio/drv/imu/imu_lsm6ds3tr_c_stm32.c#L430 with something like
> 
> ```c
> uint32_t pbdrv_imu_get_acceleration_offset(pbdrv_imu_dev_t *imu_dev) {
>     uint32_t offset = 0;
>     struct pt pt;
>     PT_INIT(&pt);
>     lsm6ds3tr_c_xl_usr_offset_get(&pt, &imu_dev->ctx, (uint8_t*)&offset);
>     lsm6ds3tr_c_xl_offset_weight_get(&pt, &imu_dev->ctx, (lsm6ds3tr_c_usr_off_w_t*)((uint8_t*)&offset) + 3);
>     return offset;
> }
> 
> void pbdrv_imu_set_acceleration_offset(pbdrv_imu_dev_t *imu_dev, uint32_t offset) {
>     struct pt pt;
>     PT_INIT(&pt);
>     lsm6ds3tr_c_xl_offset_weight_set(&pt, &imu_dev->ctx, (lsm6ds3tr_c_usr_off_w_t)(offset & 0x01000000));
>     lsm6ds3tr_c_xl_usr_offset_set(&pt, &imu_dev->ctx, (uint8_t*)&offset);
> }
> ```
> 
> I have not found motivation to test this further because I do not need that much accuracy, but I would be interested to see the results and help if someone else wants to.

Will check it when it comes, thanks for the advise!

**Reply by vanczakp** [reply]
@laurensvalk please check the team@pybricks.com mailbox. I sent you my version of 42146 controls. I would like to see what you think!

**Reply by vanczakp** [reply]
@Debenben would you please share the angle calculation of yours you mentioned (1/10 precise)? I would really appreciate it.
I placed a technic hub into the main boom (for ballast and the tilt sensor does not have accelerometer).
Thanks in advance!

> I would recommend:
> 
> * Use `acceleration` and calculate the angles yourself, that way you are not affected by gyro drift like `rotation()` and you are not limited to integer precision like `tilt()`
> * Calibrate the accelerometer with an offset e.g. somthing like [[Feature] User IMU calibration ideas #943 (comment)](https://github.com/pybricks/support/issues/943#issuecomment-1510481180)
> * Take an average of at least 10 accelerometer readings
> 
> The above should get you to 1/10 deg accuracy without problem, I checked this by attaching laser pointers to a technic hub.
> 
> * In case you need more precision you could calibrate it more precisely, e.g. like described here https://www.nxp.com/docs/en/application-note/AN4399.pdf
> * In case you need faster reaction time or want to compensate for different temperatures or things like that and you are not afraid of writing some C functions and compiling your own firmware, you should be able to e.g. set a low-pass filter on the accelerometer on a hardware level. If I am not mistaken you can also add a calibration offset here https://github.com/pybricks/pybricks-micropython/blob/master/lib/pbio/drv/imu/imu_lsm6ds3tr_c_stm32.c#L430 with something like
> 
> ```c
> uint32_t pbdrv_imu_get_acceleration_offset(pbdrv_imu_dev_t *imu_dev) {
>     uint32_t offset = 0;
>     struct pt pt;
>     PT_INIT(&pt);
>     lsm6ds3tr_c_xl_usr_offset_get(&pt, &imu_dev->ctx, (uint8_t*)&offset);
>     lsm6ds3tr_c_xl_offset_weight_get(&pt, &imu_dev->ctx, (lsm6ds3tr_c_usr_off_w_t*)((uint8_t*)&offset) + 3);
>     return offset;
> }
> 
> void pbdrv_imu_set_acceleration_offset(pbdrv_imu_dev_t *imu_dev, uint32_t offset) {
>     struct pt pt;
>     PT_INIT(&pt);
>     lsm6ds3tr_c_xl_offset_weight_set(&pt, &imu_dev->ctx, (lsm6ds3tr_c_usr_off_w_t)(offset & 0x01000000));
>     lsm6ds3tr_c_xl_usr_offset_set(&pt, &imu_dev->ctx, (uint8_t*)&offset);
> }
> ```
> 
> I have not found motivation to test this further because I do not need that much accuracy, but I would be interested to see the results and help if someone else wants to.

**Reply by Debenben** [reply]
@vanczakp When I measured the accuracy I did the calculations mainly on my pc and used the hub only for recording the values, and currently I do not have a hub I can use for testing, but you should be able to achieve 1/10 deg accuracy with something like this:


```python
from pybricks.hubs import TechnicHub
from pybricks.geometry import vector
from umath import atan2, sqrt, pi

hub = TechnicHub()

# Replace the calibration below with your own offset values obtained by this method
# https://github.com/pybricks/support/issues/943#issuecomment-1510481180
# OR instead of this calibrate use this experimental firmware
# https://github.com/pybricks/pybricks-micropython/actions/runs/11439400592
# with this calibration script
# https://github.com/pybricks/support/issues/943#issuecomment-2426464284
# and simply call hub.imu.acceleration() to get calibrated values
calibrationOffset = vector(166.89429952, -67.42990329, 236.65739606)
def get_calibrated_acceleration():
   return hub.imu.acceleration() - calibrationOffset


numberOfSamples = 10
accelerations = [[0.0, 0.0, 0.0]]*numberOfSamples
while True:
    accelerations.append(get_calibrated_acceleration())
    accelerations.pop(0)
    averageAcceleration = list(sum(accelerations[i][j]/numberOfSamples for i in range(numberOfSamples)) for j in range(3))
    x = averageAcceleration[0]
    y = averageAcceleration[1]
    z = averageAcceleration[2]
    pitch = atan2(-x, sqrt(z*z + y*y))
    roll = atan2(y, z)
    print("pitch is", pitch*180/pi, "degrees")
    print("roll is", roll*180/pi, "degrees")
```

**Reply by Debenben** [reply]
I saw you were using blocks for the crane. I have never used it but own a license so I took the opportunity to try it out for the first time to see if I could rewrite the above in blocks and it is indeed possible like this:
![pybricks_program_2024-10-22T20_08_05](https://github.com/user-attachments/assets/4e8ac258-37e3-4f7c-a0c6-1e534daeac2c)

**Reply by vanczakp** [reply]
@Debenben Wow, thank you very much! Will try it in the next few days!
This would be the new deliquent: LR12500 with double wide main boom
![462857305_8390611070976107_470027388478368577_n](https://github.com/user-attachments/assets/aa5ed79e-81c3-4cf0-8f06-d97e75e962c5)

**Reply by laurensvalk** [reply]
[Here is a firmware version](https://nightly.link/pybricks/pybricks-micropython/actions/artifacts/2103440203.zip) that uses the IMU to obtain tilt, and provides them to you as floating point values. You can install it [using these instructions](https://pybricks.com/learn/getting-started/install-pybricks/#trying-the-nightly-version-optional), but using the provided file instead.

**Reply by vanczakp** [reply]
Hello @laurensvalk! Thanks a lot! The tilt sensor is on the way, will be tested next week along with the other calculation method!
You are the best!

**Reply by laurensvalk** [reply]
What is the tilt sensor for in your application?

The firmware update suggested here should give you all the accuracy you need without any change in your own code or hardware.

**Reply by vanczakp** [reply]
@laurensvalk the tilt sensor goes into the rear boom. I need no know the angle of it to run the actuators and cable in sync and to set up the ballast height depending on the angle. The main boom has a technic hub built in (double wide boom has plenty space in it) so there I would like to calculate the angle from accelerometer. But for the derrickboom there is only space for tilt sensor which does not have accelerometer. :)

**Reply by vanczakp** [reply]
hello @laurensvalk ! Tested the special firmware. It works, but takes around 500s to have a stable value after each program start. Angle is around 58.1 deg, but after each restart it starts around 57,6-57,8 then goes up to 58.14 then back to stable 58.1. Program was triggered from computer so the crane was completely static. I would be ok with 1 min but not 8-9. 
The hub has been running for 30 min so I assume it reached its working temp. 
Fingers crossed for the plug in tilt sensor is going to behave differently and better.
Its time to try the other solution for main boom.

**Reply by laurensvalk** [reply]
If I read your post right, rounded to whole degrees the angle was 58 the entire time. Sounds very accurate.

Please note that the new method is going to be _way_ better than anything based on acceleration, including the tilt sensor.

The new method is much less noisy than measurements based on acceleration.

For example, put the hub flat on the table and have it print tilt values. Now shift it around on your table. In the new method, it correctly stays close to zero. Whereas in the case of the old acceleration based tilt it will jump around rapidly.

The same is true if you turned on any motors in your crane or moved it around slightly.

**Reply by vanczakp** [reply]
@laurensvalk what do you call new method? The upcoming calibration of hub?
Anyway I still need 0.1deg accuracy to have precise boom control combined with self balancing (moving the derrick boom paralelly to keep the balance all the time and the range for this would be 30degrees (from80-50) and it does matter if it is 30 or 300 step. I'm hoping this is possible with the actual lego hardware
Do you know why the pitch value takes so long time to get it stable in 0.1 accuracy.

**Reply by laurensvalk** [reply]
This is as stable as it gets, I think. In a giant LEGO crane with plastic components that bend, I don't think the 0.1 degree resolution is going to be your biggest issue? And keep in mind that you do get the 300 gradual steps, and even finer than that.

**Reply by vanczakp** [reply]
Hi @laurensvalk! I did an experiment today with your firmware. Hub alone laying on the table and in the main boom (motors were not loaded into the program) in two positions. While in the crane I let the cables relax for a minute and only the main boom was assembled (jib not).
As you can see while startup it takes time to get stable value in the accuracy range of 0.1. Tried with stopped program then restart immediately as well. See the results below. I think It meets my demand. Waiting one minute is not a big deal, and I dont think I would be able to see if it drifts 0,1-0,2deg. Anyway the crane is unexpectedly massive and stable! The double wide main boom did it as well as the massive and over designed superstructure. I have plans to make the turntable mostly tilting free as well: plastic spacer ring sheet into the groove of bearing. That will give me the required tension
When it works fully, I would be glad to share the 3D of crane and the program with you in case you want to build it for demonstartion. I think this is going to be the most advanced Lego crane in the world regarding functionality.
Thanks for all the help!
https://docs.google.com/spreadsheets/d/1v4l0lRfqEOR7rDUWQofbj3GrsuDYBqkFNNUddF-AqCg/edit?usp=sharing

**Reply by vanczakp** [reply]
> I saw you were using blocks for the crane. I have never used it but own a license so I took the opportunity to try it out for the first time to see if I could rewrite the above in blocks and it is indeed possible like this: ![pybricks_program_2024-10-22T20_08_05](https://private-user-images.githubusercontent.com/35615783/379004166-4e8ac258-37e3-4f7c-a0c6-1e534daeac2c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxMzM4MTUsIm5iZiI6MTczMDEzMzUxNSwicGF0aCI6Ii8zNTYxNTc4My8zNzkwMDQxNjYtNGU4YWMyNTgtMzdlMy00ZjdjLWEwYzYtMWU1MzRkYWVhYzJjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI4VDE2MzgzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTJkMDNiMzc5NDFmOGVjZGNkZWZhNDJiZTRkOThlMGE3MGZhNzU0NDJkZWQ3ZjFhMDdlZjFjNTIxNzk4NGY4NWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.88OSo2QoDMS0aIBRYTcEXB5FToU2AT_WygtlEL2m0lQ)

Hi @Debenben, I cant recreate the blocks of yours. In the for position range... I cant add the position after get #. How did you do it?
![image](https://github.com/user-attachments/assets/4852ee85-09a6-4b7b-b119-d5581b6e9de6)

**Reply by laurensvalk** [reply]
@vanczakp , @Debenben , you can drag the orange position block (the loop variable) to where you want it.

**Reply by vanczakp** [reply]
Thanks!

**Comment by Debenben** [comment]
> Hi @Debenben, I cant recreate the blocks of yours. In the for position range... I cant add the position after get #. How did you do it?

   Hmm, this ist actually the one point where I also struggled. In the end I just copied the block from tic-tac-toe example.

**Comment by Debenben** [comment]
@vanczakp Sorry for taking some time to answer. Did it work as expected?
* I tested the new firmware and results look quite good.
* I saw there was an error in my python code, I fixed it in the code above by adding `list()`
* In case you want to average, I would recommend to put the value aquisition in its own task without sleep and let it run as fast as possible and only sleep in print function. In case that is still relevant I can show you in more detail.
* Not sure if you noticed, but I thought I'd mention it: When you position the hub in the crane, make sure the angle of the main boom is the "roll" angle. The convention pybricks uses is like that on a plane: first heading, then pitch, then roll. You want to first apply the angle that the crane is tilted sideways due to it standing on a slope, this is therefore "pitch". Afterwards you apply the "roll" as position of the boom around the mechanical axis which is pointing along this slope. If you mistakenly take "pitch" as the angle of the boom, the roll angle becomes the rotation of the boom along the booms forward axis. If you do it the wrong way your roll angle gets unstable when the boom is pointing up since it is then your heading which you cannot measure with gravity. If you do it correctly the same happens to your roll angle (the boom position) when the crane has fallen over and is lying sideways, but I guess that is not a use-case.
* The imu has an integrated temperature sensor which would require changes in firmware in order to read out. I do not think it is worth the effort but I thougth I'd mention it in case temperature is a big problem for you.
* The external tilt sensor from lego unfortunately only outputs integers. A little improvement is possible by using `sensor.read(3)` instead of `sensor.tilt()` to get full 360 degree instead of just -45 to 45.

@laurensvalk I guess my problem was that I tried dragging the position block when the text input was active (I guess it is only dragging text and not the block in that case). The text-selection mouse cursor when hovering over the input box is a bit misleading because you first have to click in order to select text, I think it would have helped me if it was dragging or selecting mouse cursor instead. In my opinion one could even hide the functionality to rename the iterator behind double-click or long-press since renaming is some esthetic improvement useful in nested loops while getting the value is essential.

**Reply by vanczakp** [reply]
hello @laurensvalk ! No problem.

- It did work as expected, but the tilt sensor in the technic hub is more accurate. After 5 minutes when there is no drift anymore the tilt sensor's value range was around 0.05 (min max of few minutes measurement with one sec sampling rate) compared to the accelerometer's 0,1 range.
- regarding the roll pitch: I have to check it If I get it right but I get what it should mean and I have to check if I did right.
- The external tilt sensor accuracy is a pity. I wanted it to use for calibrating the actuator and the spanner cable of derrick boom when start up. I have the form to link the actuator movement to the cable but it depends on the angle a lot and 1deg is not enough, On the other hand the value I get from it is way too unstable when the boom is moving, it can jump +/-10 degrees easily then back... I'll do the floating ballst version with moving derrick boom but first the V-frame which requires static derrick boom.

I spent the last few weeks to make the crane more robust, I think I reached the max I what someone can achieve. Main problem is when the crane is in balance, standing on a completely flat ground, and I rotate the turntable, I can measure the imperfection of the bearing via getting different angles from the superstructures. I realized my floor is not close to flat at all (some area has 0,8deg slope). Rotating the crane on slope gives you even more strange angles as the turntable tries to push back the hole structure to balance. I must do a DOE and see the and the impact of different slopes on the measured angle (two hubs (90deg rotated to each other) measures the floor angle )
On the other hand I just realized If I change the ballast weight then I would need different gain for the control curcuit. So I just questioned myself if it worth it or not. Its getting too much, regardless I have all the knowledge to do it.
